text/0000-mut-pattern-shorthand.md

```diff
@@ -0,0 +1,569 @@
+let Person { age, hobby, pocket: [mut x, mut y, mut z] }
+  = Person { age: 1, hobby: "designing Rust"
+           , pocket: ["key", "phone", "headphones"] };
+```
+
+After:
+
+```rust
+let mut (x, y) = (1, 2);
+let (mut (x, y), z) = ((1, 2), 3);
+let mut ((x, y), z) = ((1, 2), 3);
+
+struct Person<'a> { age: usize, hobby: &'a str, pocket: [&'a str; 3] }
+
+let Person { age, hobby, pocket: mut [x, y, z] } 
+  = Person { age: 1, hobby: "designing Rust"
+           , pocket: ["key", "phone", "headphones"] };
+```
+
+## `if let`
+
+Before:
+
+```rust
+if let Ok((mut x, mut y)) = Ok((1, 2)) { ... }
+```
+
+After:
+
+```rust
+if let Ok(mut (x, y)) = Ok((1, 2)) { ... }
+```
+
+## `while let`
+
+Before:
+
+```rust
+let mut arr = vec![(1, 2), (2, 3), (3, 4)];
+let mut iter = arr.drain(..);
+
+while let Some((mut x, mut y)) = iter.next() { ... }
+```
+
+After:
+
+```rust
+let mut arr = vec![(1, 2), (2, 3), (3, 4)];
+let mut iter = arr.drain(..);
+
+while let Some(mut (x, y)) = iter.next() { ... }
+```
+
+## `for` loops
+
+Before:
+
+```rust
+let mut arr = vec![(1, 2), (2, 3), (3, 4)];
+for (mut x, mut y) in arr.drain(..) { ... }
+```
+
+After:
+
+```rust
+let mut arr = vec![(1, 2), (2, 3), (3, 4)];
+for mut (x, y) in arr.drain(..) { ... }
+```
+
+## `match` expressions
+
+Before:
+
+```rust
+match (1, 2) {
+    (mut x, mut y) => {}
+}
+```
+
+After:
+
+```rust
+match (1, 2) {
+    mut (x, y) => {}
+}
+```
+
+## `fn` arguments
+
+Before:
+
+```rust
+fn do_stuff((mut x, mut y): (usize, usize)) { ... }
+```
+
+After:
+
+```rust
+fn do_stuff(mut (x, y): (usize, usize)) { ... }
+```
+
+## Lambdas
+
+Before:
+
+```rust
+let lam = |(mut x, mut y): (u8, u8)| { x = x + 1; (x, y) };
+```
+
+After:
+
+```rust
+let lam = |mut (x, y): (u8, u8)| { x = x + 1; (x, y) };
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+We modify the grammar by adding, to [`pat`](https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y#L958)
+the following productions:
+
+```rust
+pat
+: // other existing productions
+| MUT '(' pat_tup ')'
+| MUT '[' pat_vec ']'
+```
+
+Alternatively, the change can be seen as adding `MUT?` to the referenced
+productions above.
+
+Note that `let mut mut mut x = 4;` is not a legal production.
+A `mut` may not be immediately followed by another `mut`.
+
+## Semantics
+
+The semantics of a `mut PAT` pattern is defined as the semantics of the
+desugared form where any binding inside `PAT` introduced is a `mut` binding.
+Consult the [guide level explanation][guide-level-explanation] for some examples
+of desugarings.
+
+## Warnings
+
+The following snippet `let mut (mut x, mut y) = ...;`, while grammatically valid,
+also has redundant `mut`s either outside the tuple or inside the tuple.
+When such a redundancy happens, the compiler should unconditionally warn the
+user. Furthermore, when `let mut (x, y)` results in partial unused `mut`ability
+then the compiler should also warn.
```

Comment:
nitpick: I don't think "unconditionally" is appropriate; this can just be an extension of `unused_mut` (or maybe a new lint, but distinguishing seems unnecessary to me).

So you get something suppressible; maybe
```rust
warning: variable is already mutable
 --> src/main.rs:2:9
  |
2 |     let mut (mut x, y) = (4, 5;
  |              ----^
  |              |
  |              help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default
```

Reply:
Yeah; I think your comment reflects what I actually meant ;)